[IMP] website_event_booth{_exhibitor}{_sale}: add chevrons to booth process

Odoonan opened the pull request:
CHEVRONS

Add chevrons to the event booth front-end process. A template is added
to be called in booth choice and form views, and the one called from js
is also updated with chevrons. The code is minimized, and fits for the
three current steps only.
- Booth Selection
- Contact Details
- Confirmed / Payment

CLICKABLE + REMEMBER BOOTH CHOICE

To ease the process of changing booth / booking another one, the
"Booth Selection" chevron is clickable and highlighted on hovering.
Also, it will keep booth_ids and booth_category_id url parameters.
From that, the category will be selected and the booth_ids checked
(if available), the first time the selection screen is loaded (once
the user goes between booth categories, it will not remember booths
any more and not preselect any) In order to do so, the arguments are
added to the appropriate methods in controller (and overrides), with
particular attention to match format. (string vs int...)

SCSS

Styling is added in website_event_booth.scss. Cautions are taken to
ensure proper chevrons coloring even with edited theme colors.

DYNAMIC WORDING

Instead of 'confirmed', 'payment' is used in chevrons, and "Go to
payment" instead of "Book my Booths" in the contact details form,
if one of the following is true (~ new boolean value 'use_payment_step'):
- In selection screen: at least one category is not free
- In Contact Details screen: the chosen booth is not free
- In both Selection and Contact screens: if cart has non null total
amount. (this is the current logic used to redirect to payment)

CONTROLLER REORGANIZATION

In order to add the "use_payment_step" to the rendering values,
the routes of those views now have their value preparation methods,
making them easily overriden (done while maintaining guidelines)

OTHER CHANGES

- Some wording on the form
- Hide sponsors in the booth booking flow
- All views have the same space between chevrons and content (48px)
- Tour is updated to match new wording.
- New map icon and simple link button to display the map of booths.
- No more justify content for display of booth categories
- unglue the rows of booths vertically


Task-2821962

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/114052).

thiryr commented on addons/website_event_booth/static/src/scss/website_event_booth.scss:
```diff
@@ -71,3 +71,53 @@ label.o_wbooth_category_unavailable {
+        padding: 0 24px 0 48px;
+        height: 48px;
+        line-height: 48px;
+        width: 33.333%;
```
```suggestion
        width: calc(100%/3);
```

There's still some rounding non-sense going on which makes the last element look like it has 2 borders at some specific width.
Maybe we can use a css grid but it's arguably less supported and I'm not sure it would solve the issue, leaving it up to you :)

thiryr reviewed (commented):
Hey :waffle: 
Not much to say, pretty clean imp :)

A couple nitpicks and it's good for me

Odoonan reviewed (commented):
Hello @thiryr , Thanks for review :) Changes done and a small comment :)

thiryr commented:
Oh yeah, 33 33 34 is a good fix ^^

awa-odoo commented on addons/website_event_booth_sale/controllers/event_booth.py:
```diff
@@ -9,12 +9,24 @@
 
 class WebsiteEventBoothController(WebsiteEventController):
 
+    def _prepare_booth_contact_form_values(self, event, booth_ids, booth_category_id):
```
We usually put all routes together and then all utility methods, sorted alphabetically :socks:

awa-odoo commented on addons/website_event_booth_sale/controllers/event_booth.py:
```diff
@@ -9,12 +9,24 @@
 
+    def _prepare_booth_contact_form_values(self, event, booth_ids, booth_category_id):
+        values = super()._prepare_booth_contact_form_values(event, booth_ids, booth_category_id)
+        values['use_payment_step'] = request.website.sale_get_order().amount_total or \
```
Maybe "has_payment_step" ?

awa-odoo reviewed (commented):
Hello @Odoonan 

Cool little feature :slightly_smiling_face: 
Left some comments here and there but nothing major.

Sorry for the review delay.

Cheers!

Odoonan commented on addons/website_event_booth/static/src/js/booth_register.js:
```diff
@@ -57,16 +66,20 @@ publicWidget.registry.boothRegistration = publicWidget.Widget.extend({
         var $boothElem = this.$('.o_wbooth_booths');
         $boothElem.empty();
+        var self = this;
         $.each(this.boothCache[this.activeType], function (key, booth) {
```
I'm having a look :)

Odoonan commented on addons/website_event_booth/static/src/scss/website_event_booth.scss:
```diff
@@ -71,3 +71,54 @@ label.o_wbooth_category_unavailable {
+        padding-left: 24px;
+        height: 48px;
+        line-height: 48px;
+        width: 33%;
```
If we want to avoid any rounding issue for high res screens and have an unique size, we could go with a simpler calc(100%/3) , as it seems it is used in other places, actually.

Odoonan commented on addons/website_event_booth/static/src/scss/website_event_booth.scss:
```diff
@@ -71,3 +71,54 @@ label.o_wbooth_category_unavailable {
+            width: 34%;
+            background: $wevent-booth-progress-bar-color;
+            .chevron:before {
+                border-left: 24px solid $wevent-booth-progress-bar-color;
```
Yes that is also needed to have correctly colored buttons on hover with those shapes. It is basically of I understood the chevrons worked. I had tried other ideas at the time but did not find other way to do so. Same for JEH at the time :)

Odoonan commented on addons/website_event_booth/static/src/scss/website_event_booth.scss:
```diff
@@ -71,3 +71,54 @@ label.o_wbooth_category_unavailable {
+            border-right: 0;
+            border-left: 24px solid $wevent-booth-progress-bar-color;
+            &:before {
+                position: absolute;
```
I did, from my previous work on appointment. Actually chevrons are quite tricky. Here, the 'border' is indeed as large as the chevron. (or 'is' the chevron, you could say) You should be able to play on the runbot now ^^

Odoonan commented on addons/website_event_booth/static/src/scss/website_event_booth.scss:
```diff
@@ -71,3 +71,54 @@ label.o_wbooth_category_unavailable {
+        &:first-child {
+            padding-left: 0px;
+            &:hover:not(.o_current_stage) {
+                background-color: $wevent-booth-progress-bar-color-2 !important;
```
Yes, as once at the confirmation, we do not want to allow coming back to details since the selected booths will be booked ^^'

Odoonan commented on addons/website_event_booth/views/event_booth_templates.xml:
```diff
@@ -120,4 +123,20 @@
+            <li t-attf-class="position-relative float-start m-0 text-center #{'o_current_stage' if step=='STEP_DETAILS_FORM' else 'text-muted'}">
+                Contact Details<span class="chevron position-absolute d-block"/>
+            </li>
+            <li class="position-relative float-start m-0 text-center text-muted">
```
Not in this template. There is a qweb template called from the JS for that case (free + confirmed). (event_booth_registration_complete, only for the confirmation stage, where o_current_stage is indeed used). If not free, we go to the sale process instead of first stage. The one here is used for the other cases.

Odoonan commented on addons/website_event_booth/static/src/js/booth_register.js:
```diff
@@ -54,18 +61,22 @@ publicWidget.registry.boothRegistration = publicWidget.Widget.extend({
+        let self = this;
+        for (const booth of this.boothCache[this.activeType]) {
+            let selectedBooth = self.boothsFirstRendering && self.selectedBoothIds.includes(booth.id.toString());
             let $checkbox = dom.renderCheckbox({
```
Here, renderCheckbox returns a jquery object, so I kept a few while we keep this solution.

Notably .append, maybe we could use insertAdjacentHTML and .outerHTML but while we have it...

Also I am not sure what the recommended solution is when it comes to replacing an element with the result of a Qweb.render

Odoonan reviewed (commented):
Hello @awa-odoo ,

I've done most of the changes, removed jquery (partially) in the methods I worked in, and changed the structure a bit. I've left a few comments if you want to have a look.
(Did not squash commits yet)

You should be able to give it a go on the runbot, too ^^

Cheers!

awa-odoo commented on addons/website_event_booth/static/src/js/booth_register.js:
```diff
@@ -213,7 +224,12 @@ publicWidget.registry.boothRegistration = publicWidget.Widget.extend({
+                const orderProgressBarEl = this.el.querySelector('.o_wevent_booth_order_progress');
+                orderProgressBarEl.parentElement.removeChild(orderProgressBarEl);
+                const boothCategoryId = this.el.querySelector('input[name=booth_category_id]').value;
                 $form.replaceWith($(QWeb.render('event_booth_registration_complete', {
```
`$form[0].replaceWith(QWeb.render('event_booth_registration_complete', { ... }));` would work?

#removejquery

awa-odoo commented on addons/website_event_booth/static/src/xml/event_booth_registration_templates.xml:
```diff
@@ -1,6 +1,19 @@
 <templates>
 
     <t t-name="event_booth_registration_complete">
+        <ul class="o_wevent_booth_order_progress list-unstyled d-none d-lg-block mb-0">
```
Can't we t-call "event_booth_order_progress" ? Seems sad to duplicate the code.

awa-odoo commented on addons/website_event_booth/static/src/js/booth_register.js:
```diff
@@ -16,11 +16,18 @@ publicWidget.registry.boothRegistration = publicWidget.Widget.extend({
+            if (boothSelectedCategory) {
+                // A single id is interpreted as a number. Hence the toString method.
+                this.selectedBoothIds = this.el.querySelector('.o_wbooth_booths').dataset.selectedBoothIds.toString().split(',');
+                boothSelectedCategory.click();
```
I see that we still have some post-processing, is that part complex to do in the server-rendered template?

awa-odoo commented on addons/website_event_booth/static/src/js/booth_register.js:
```diff
@@ -54,18 +61,22 @@ publicWidget.registry.boothRegistration = publicWidget.Widget.extend({
-        var $boothElem = this.$('.o_wbooth_booths');
-        $boothElem.empty();
-        $.each(this.boothCache[this.activeType], function (key, booth) {
+        let boothsElem = this.el.querySelector('.o_wbooth_booths');
+        boothsElem.innerHTML = '';
+        let self = this;
```
Those don't change so why "let" ?

awa-odoo commented on addons/website_event_booth/static/src/js/booth_register.js:
```diff
@@ -213,7 +224,12 @@ publicWidget.registry.boothRegistration = publicWidget.Widget.extend({
 
             const jsonResponse = response && JSON.parse(response);
             if (jsonResponse.success) {
+                const orderProgressBarEl = this.el.querySelector('.o_wevent_booth_order_progress');
+                orderProgressBarEl.parentElement.removeChild(orderProgressBarEl);
+                const boothCategoryId = this.el.querySelector('input[name=booth_category_id]').value;
```
I guess it's easier to re-render the whole thing rather than just "updating" the existing one?

Would have been easier with OWL and re-active stuff :eyes: 

Also, isn't `orderProgressBarEl.parentElement.removeChild(orderProgressBarEl);` the same as juste `orderProgressBarEl.remove()` ?

awa-odoo reviewed (commented):
Hello @Odoonan 

A bit easier to understand once you have a runbot to play with, it looks nice!

Had a complete re-read and left some additional comments and questions.

Sorry for the delay :cold_face: and cheers!

awa-odoo commented:
@Odoonan side-note: there is a small file conflict to resolve :)

Odoonan commented:
Hello @awa-odoo .

I updated the code and restructured a bit the JS. I left a few of your comments open for inspection.

Just a quick summary of how this code is atm for the three stages: booth selection, contact form, confirmation. The three stages use the boothRegistration widget.

- The route /booth will render the booth registration. Everything is rendered except the list of booth checkboxes. As the template is called from JS on booth category change, it is only rendered once we are in the start method of the widget.
- On confirmation of the selection, we go to /register route and the rendered form is returned
- On confirmation of the form, we stay on the same page and the form is replaced from JS by confirmation screen.

We could maybe move the logic (and add the xml) of pre-selected checkboxes to the first rendering of the event_booth_registration template rendered by /booth. But it also means that some of that template will be only used at loading and replaced from JS when changing category with the new template event_booth_checkbox_list. It would simplify the preprocessing and remove the "boothFirstRendering" variable though. What do you think?

Hopefully this is clear. Thanks for your input :) Feel free to shout if something feels wrong.

Cheers!

awa-odoo reviewed (commented):
Hello @Odoonan 

A few last details to check.

And then, as discussed, we will keep the approach with 2 templates as it makes the transition from the existing code a lot easier.

Cheers!

Odoonan commented on addons/website_event_booth/static/src/js/booth_register.js:
```diff
@@ -213,10 +218,14 @@ publicWidget.registry.boothRegistration = publicWidget.Widget.extend({
             if (jsonResponse.success) {
-                $form.replaceWith($(QWeb.render('event_booth_registration_complete', {
+                this.el.querySelector('.o_wevent_booth_order_progress').remove();
+                const boothCategoryId = this.el.querySelector('input[name=booth_category_id]').value;
```
yes,the value of category of booked booths when submitting the form. activeType is the 'selected' category on the selection page. I renamed activeType to activeBoothCategoryId

tde-banana-odoo commented:
@robodoo override=ci/security

Reason: replaced old jQuery DOM building with native "innerHTML" call. Inserted HTML is a small QWeb rendered template where we control content and therefore output.

awa-odoo commented:
@robodoo r+
